File: app/src/main/java/rasel/lunar/launcher/home/HomeUtils.java
/*
 * Lunar Launcher
 * Copyright (C) 2022 Md Rasel Hossain
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package rasel.lunar.launcher.home;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.text.format.DateFormat;
import android.view.View;

import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import java.util.Calendar;

import rasel.lunar.launcher.R;
import rasel.lunar.launcher.helpers.Constants;
import rasel.lunar.launcher.helpers.SwipeTouchListener;
import rasel.lunar.launcher.helpers.UniUtils;
import rasel.lunar.launcher.qaccess.QuickAccess;
import rasel.lunar.launcher.settings.SettingsActivity;
import rasel.lunar.launcher.todos.TodoManager;

public class HomeUtils {

    Constants constants = new Constants();
    UniUtils uniUtils = new UniUtils();

    // Gets time format
    protected String getTimeFormat(SharedPreferences sharedPreferences, Context context) {
        int timeFormatValue = sharedPreferences.getInt(constants.SHARED_PREF_TIME_FORMAT, 0);
        String timeFormat = "";
        switch (timeFormatValue) {
            case 0:
                if(DateFormat.is24HourFormat(context)) {
                    timeFormat = "kk:mm";
                } else {
                    timeFormat = "h:mm a";
                }
                break;
            case 1: timeFormat = "h:mm a";
                break;
            case 2: timeFormat = "kk:mm";
                break;
        }
        return timeFormat;
    }

    // Gets number suffixes
    private String getDateNumberSuffix(){
        Calendar calendar = Calendar.getInstance();
        int date = calendar.get(Calendar.DAY_OF_MONTH);
        switch (date) {
            case 1:     case 21:    case 31:
                return "ˢᵗ";
            case 2:     case 22:
                return "ⁿᵈ";
            case 3:     case 23:
                return "ʳᵈ";
            default: return "ᵗʰ";
        }
    }

    // Gets date format
    protected String getDateFormat(SharedPreferences sharedPreferences) {
        int dateFormatValue = sharedPreferences.getInt(constants.SHARED_PREF_SHOW_YEAR, 1);
        String dateFormat = "";
        switch (dateFormatValue) {
            case 0: dateFormat = "EEE',' d" + getDateNumberSuffix() + " MMMM";
                break;
            case 1: dateFormat = "EEE',' d" + getDateNumberSuffix() + " MMM',' yyyy";
                break;
        }
        return dateFormat;
    }

    // Gestures on root view
    protected void rootViewGestures(View view, Context context, FragmentManager fragmentManager, FragmentActivity fragmentActivity, int lockMethodValue) {
        view.setOnTouchListener(new SwipeTouchListener(context) {
            @Override
            public void onSwipeUp() {
                super.onSwipeUp();
                (new QuickAccess()).show(fragmentManager, constants.MODAL_BOTTOM_SHEET_TAG);
            }
            @Override
            public void onSwipeDown() {
                super.onSwipeDown();
                uniUtils.expandNotificationPanel(context, fragmentActivity);
            }
            @Override
            public void onDoubleClick() {
                super.onDoubleClick();
                uniUtils.lockMethod(lockMethodValue, context, fragmentActivity);
            }
        });
    }

    // Gestures on battery progress indicator
    protected void batteryProgressGestures(View view, Context context, FragmentActivity fragmentActivity, int lockMethodValue) {
        view.setOnTouchListener(new SwipeTouchListener(context) {
            @Override
            public void onLongClick() {
                super.onLongClick();
                fragmentActivity.startActivity(new Intent(context, SettingsActivity.class));
            }
            @Override
            public void onSwipeDown() {
                super.onSwipeDown();
                uniUtils.expandNotificationPanel(context, fragmentActivity);
            }
            @Override
            public void onDoubleClick() {
                super.onDoubleClick();
                uniUtils.lockMethod(lockMethodValue, context, fragmentActivity);
            }
        });
    }

    // Gestures on todos section
    protected void todosGestures(View view, Context context, FragmentManager fragmentManager, FragmentActivity fragmentActivity, int lockMethodValue) {
        view.setOnTouchListener(new SwipeTouchListener(context) {
            @Override
            public void onLongClick() {
                super.onLongClick();
                (fragmentManager.beginTransaction()).add(R.id.main_fragments_container,
                        new TodoManager()).addToBackStack("").commit();
            }
            @Override
            public void onSwipeUp() {
                super.onSwipeUp();
                (new QuickAccess()).show(fragmentManager, constants.MODAL_BOTTOM_SHEET_TAG);
            }
            @Override
            public void onSwipeDown() {
                super.onSwipeDown();
                uniUtils.expandNotificationPanel(context, fragmentActivity);
            }
            @Override
            public void onDoubleClick() {
                super.onDoubleClick();
                uniUtils.lockMethod(lockMethodValue, context, fragmentActivity);
            }
        });
    }
}

File: app/src/main/kotlin/rasel/lunar/launcher/home/HomeUtils.kt
/*
 * Lunar Launcher
 * Copyright (C) 2022 Md Rasel Hossain
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package rasel.lunar.launcher.home

import android.content.Intent
import android.content.SharedPreferences
import android.text.format.DateFormat
import android.view.View
import android.widget.Toast
import androidx.biometric.BiometricPrompt
import androidx.fragment.app.FragmentActivity
import rasel.lunar.launcher.R
import rasel.lunar.launcher.helpers.Constants
import rasel.lunar.launcher.helpers.SwipeTouchListener
import rasel.lunar.launcher.helpers.UniUtils
import rasel.lunar.launcher.qaccess.QuickAccess
import rasel.lunar.launcher.settings.SettingsActivity
import rasel.lunar.launcher.todos.TodoManager
import java.util.*


internal class HomeUtils(
    private val fragmentActivity: FragmentActivity,
    private val sharedPreferences: SharedPreferences) {

    private val fragmentManager = fragmentActivity.supportFragmentManager
    private val context = fragmentActivity.applicationContext
    private val constants = Constants()
    private val uniUtils = UniUtils()
    private val quickAccess = QuickAccess()

    /* get time format string */
    val timeFormat: String? get() {
        when (sharedPreferences.getInt(constants.KEY_TIME_FORMAT, 0)) {
            0 -> return if (DateFormat.is24HourFormat(context)) {
                "kk:mm"
            } else {
                "h:mm a"
            }
            1 -> return "h:mm a"
            2 -> return "kk:mm"
        }
        return null
    }

    /* get date number suffix */
    private val dateNumberSuffix: String get() {
        val calendar = Calendar.getInstance()
        return when (calendar[Calendar.DAY_OF_MONTH]) {
            1, 21, 31 -> "ˢᵗ"
            2, 22 -> "ⁿᵈ"
            3, 23 -> "ʳᵈ"
            else -> "ᵗʰ"
        }
    }

    /* get date format string */
    val dateFormat: String get() {
        val dateFormatValue = sharedPreferences.getString(
            constants.KEY_DATE_FORMAT,
            constants.DEFAULT_DATE_FORMAT
        )
        return if (dateFormatValue!!.contains("x")) {
            dateFormatValue.replace("x", dateNumberSuffix)
        } else {
            dateFormatValue
        }
    }

    /* gestures on root view */
    fun rootViewGestures(view: View, lockMethodValue: Int) {
        view.setOnTouchListener(object : SwipeTouchListener(context) {
            /* open quick access panel on swipe up */
            override fun onSwipeUp() {
                super.onSwipeUp()
                quickAccess.show(fragmentManager, constants.BOTTOM_SHEET_TAG)
            }
            /* expand notification panel on swipe down */
            override fun onSwipeDown() {
                super.onSwipeDown()
                uniUtils.expandNotificationPanel(context)
            }
            /* lock the screen on double tap (optional) */
            override fun onDoubleClick() {
                super.onDoubleClick()
                uniUtils.lockMethod(lockMethodValue, context, fragmentActivity)
            }
        })
    }

    /* gestures on battery progress indicator area */
    fun batteryProgressGestures(view: View, lockMethodValue: Int) {
        view.setOnTouchListener(object : SwipeTouchListener(context) {
            /* open settings activity on long click */
            override fun onLongClick() {
                super.onLongClick()
                fragmentActivity.startActivity(Intent(context, SettingsActivity::class.java))
            }
            /* expand notification panel on swipe down */
            override fun onSwipeDown() {
                super.onSwipeDown()
                uniUtils.expandNotificationPanel(context)
            }
            /* lock the screen on double tap (optional) */
            override fun onDoubleClick() {
                super.onDoubleClick()
                uniUtils.lockMethod(lockMethodValue, context, fragmentActivity)
            }
        })
    }

    /* gestures on todo area */
    fun todosGestures(view: View, lockMethodValue: Int) {
        view.setOnTouchListener(object : SwipeTouchListener(context) {
            /* open TodoManager on long click */
            override fun onLongClick() {
                super.onLongClick()
                when (sharedPreferences.getBoolean(constants.KEY_TODO_LOCK, false)) {
                    false -> fragmentManager.beginTransaction().add(R.id.main_fragments_container, TodoManager())
                        .addToBackStack("").commit()
                    /* show authentication screen if lock is on */
                    true -> {
                        if (uniUtils.canAuthenticate(context)) {
                            val biometricPrompt = BiometricPrompt(fragmentActivity, authenticationCallback)
                            try {
                                biometricPrompt.authenticate(uniUtils.biometricPromptInfo(fragmentActivity.getString(R.string.todo_manager), fragmentActivity))
                            } catch (exception: Exception) {
                                exception.printStackTrace()
                            }
                        }
                    }
                }
            }
            /* open quick access panel on swipe up */
            override fun onSwipeUp() {
                super.onSwipeUp()
                quickAccess.show(fragmentManager, constants.BOTTOM_SHEET_TAG)
            }
            /* expand notification panel on swipe down */
            override fun onSwipeDown() {
                super.onSwipeDown()
                uniUtils.expandNotificationPanel(context)
            }
            /* lock the screen on double tap (optional) */
            override fun onDoubleClick() {
                super.onDoubleClick()
                uniUtils.lockMethod(lockMethodValue, context, fragmentActivity)
            }
        })
    }

    /* authentication callback for TodoManager lock */
    private val authenticationCallback = object : BiometricPrompt.AuthenticationCallback() {
        override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
            fragmentManager.beginTransaction().add(R.id.main_fragments_container, TodoManager())
                .addToBackStack("").commit()
        }
        override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
            Toast.makeText(context, fragmentActivity.getString(R.string.authentication_error), Toast.LENGTH_SHORT).show()
        }
        override fun onAuthenticationFailed() {
            Toast.makeText(context, fragmentActivity.getString(R.string.authentication_failed), Toast.LENGTH_SHORT).show()
        }
    }

}
